docs/ebpf.py

```diff
@@ -0,0 +1,229 @@
+        )
+
+
+def code_example(
+    symbol: str,
+    title: str = None,
+    language: str = "",
+    lines: bool = True,
+    signature: bool = False,
+    path: str = "",
+) -> str:
+    """
+    Include the body of a code example.
+
+    `symbol` takes the name of the function or snippet to include.
+    `title` is rendered as a title at the top of the snippet.
+    `language` is the name of the programming language passed to pygments.
+    `lines` controls rendering line numbers.
+    `signature` controls whether or not the function signature and brackets are
+        included.
+    `path` specifies the include path that may contain globs.
+    """
+    opts: List[str] = []
+    if lines:
+        opts.append("linenums='1'")
+    if title:
+        opts.append(f"title='{title}'")
+
+    if signature:
+        body = full_body(path, symbol)
+    else:
+        body = inner_body(path, symbol)
+
+    out = f"``` {language} {' '. join(opts)}\n{body}```"
+
+    return out
+
+
+def inner_body(path: str, sym: str) -> str:
+    """
+    Get the inner body of sym, using default delimiters.
+
+    First and last lines (so, function signature and closing bracket) are
+    stripped, the remaining body dedented.
+    """
+    out = _search_body(path, sym)
+    if len(out) < 2:
+        raise ValueError(
+            f"Need at least two lines to get inner body for symbol {sym}"
+        )
+
+    return textwrap.dedent("".join(out[1:-1]))
+
+
+def full_body(path: str, sym: str) -> str:
+    """Get the full body of sym, using default delimiters, dedented."""
+    out = _search_body(path, sym)
+
+    return textwrap.dedent("".join(out))
+
+
+def _get_body(
+    f: TextIOWrapper, sym: str, start: str = "{", end: str = "}"
+) -> List[str]:
+    """
+    Extract a body of text between sym and start/end delimiters.
+
+    Tailored to finding function bodies of C-family programming languages with
+    curly braces.
+
+    The starting line of the body must contain sym prefixed by a space, with
+    d_start appearing on the same line, for example " Foo() {". Further
+    occurrences of "{" and its closing counterpart "}" are tracked, and the
+    lines between and including the final "}" are returned.
+    """
+    found = False
+    stack = 0
+    lines = []
+
+    sym = " " + sym
+
+    for line in f.readlines():
+        if not found:
+            # Skip current line if we're not in a body and the current line
+            # doesn't contain a start delimiter.
+            if sym not in line:
+                continue
+
+            found = True
+
+        # Count the amount of start delimiters.
+        stack += line.count(start)
+
+        if stack == 0:
+            # No opening delimiter found, ignore the line.
+            found = False
+            continue
+
+        lines.append(line)
+
+        # Count the amount of end delimiters and stop if we've escaped the
+        # current scope.
+        stack -= line.count(end)
+        if stack <= 0:
```

Comment:
It will not include the final line in this case, and that's kind of by design. A character-based tokenizer would be more complex and we don't need that here.